src/main/java/org/folio/rest/impl/PostOrderLineHelper.java

```diff
@@ -56,34 +86,179 @@
     subObjFuts.add(createReportingCodes(compPOL, line, compPOL.getReportingCodes()));
     subObjFuts.add(createFundDistribution(compPOL, line, compPOL.getFundDistribution()));
 
-    CompletableFuture.allOf(subObjFuts.toArray(new CompletableFuture[subObjFuts.size()]))
-      .thenAccept(v -> {
+    return CompletableFuture.allOf(subObjFuts.toArray(new CompletableFuture[subObjFuts.size()]))
+      .thenCompose(v -> {
         try {
           Buffer polBuf = JsonObject.mapFrom(line).toBuffer();
-          httpClient.request(HttpMethod.POST, polBuf, resourcesPath(PO_LINES), okapiHeaders)
+          return httpClient.request(HttpMethod.POST, polBuf, resourcesPath(PO_LINES), okapiHeaders)
             .thenApply(HelperUtils::verifyAndExtractBody)
-            .thenAccept(body -> {
+            .thenApply(body -> {
               logger.info("response from /po_line: " + body.encodePrettily());
 
               compPOL.setId(body.getString("id"));
-              future.complete(compPOL);
-            })
-            .exceptionally(t -> {
-              logger.error("failed to update CompositePoLine", t);
-              future.completeExceptionally(t.getCause());
-              return null;
+              return compPOL;
             });
         } catch (Exception e) {
           logger.error("Exception calling POST /po_line", e);
-          future.completeExceptionally(e.getCause());
+          throw new CompletionException(e);
         }
+      });
+  }
+
+  /**
+   * Retrieves product type details associated with given PO line
+   * and builds 'product type name' -> 'product type id' map.
+   *
+   * @param compPOL the PO line to retrieve product type details for
+   * @return product types map
+   */
+  private CompletableFuture<Map<String, String>> getProductTypesMap(PoLine compPOL) {
+    // do not fail if no productId is provided, should be enforced on schema level if it's required
+    if (compPOL.getDetails() == null || compPOL.getDetails().getProductIds().isEmpty()) {
+      return completedFuture(Collections.emptyMap());
+    }
+
+    String endpoint = compPOL.getDetails().getProductIds().stream()
```

Comment:
Damn... I meant to bring this up sooner.  I think we're eventually going to want to store the identifier UUID and not have our own productIdType enumeration which is redundant and will inevitably become confusing at some point.  For now I think what you have here is fine, but we'll have to change this later on...  I think ideally the UI would get the list of available identifierTypes and populate a drop-down.  When making the API call, it would use the UUID, not the name/value.  This is similar to how we plan on handling things like locations, assignedTo, etc.

Reply:
Yes, we actually discussed this earlier and decided to not change schema in scope of this story and make this change separately. But in short term, implement the logic above.